### PR TITLE
Minimal changes to get pmxbot up and running again

### PR DESCRIPTION
**Pull request by chrisBrookes93**
Minimal changes to get pmxbot up and running again. Our main use case is webhooks as we've lost observability on applications as a result. This is now working as well as the common integrations (e.g. thanks and motivate)

Key changes:
- Changed the dependency to be the new name: `slack_sdk`
- Updated to use the v2 of the RTMClient
- `self.slack.server` is no longer a thing in the new client, so changed to use `self.slacker` for retrieving info on `users`/`channels` (or as they're referred to as now: `conversations`)
- Now using the web client for posting messages to slack. For this we need the channel ID rather than the channel name, so there is logic implemented to resolve this where necessary.
- Kept it simple by spawning a thread to run the scheduler rather than an asyncio loop as suggested by jaraco here: https://github.com/pmxbot/pmxbot/issues/85#issuecomment-489086969

Talking to PmxBot directly:
image

Webhooks and scheduled messages:

image

Responding in threads, tagging users and channels:

image

Help docs:

image

Some other random commands:

image

image

### CODE REVIEW

File: setup.cfg
Comment (jaraco) [inline setup.cfg:97]: This comment becomes stale

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:98]: This code looks potentially overly-conservative. Will it work the same with `.body['channel']['name']`? I would be surprised if passing `None` for the `channel_name` would work as expected.

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:80]: Why change the sleep time here? Setting to `1` would mean that scheduled events could happen up to 1 sec after they're scheduled, which could potentially create a user-visible delay.

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:148]: I'd like to refactor this bit to separate concerns (generating channel mappings, consuming channel mappings). I'll help with this.

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:69]: I wonder if `self.slack` has an event loop that could periodically invoke the scheduler rather than creating a thread. A thread may be fine, but it could also introduce threading issues (what if a message arrives at the same time as a scheduled event).

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:73]: Did you consider instead using `self.slack.on("message")(self.handle_message)` and updating `handle_message` to accept `(client, event)`?

**Review by jaraco — APPROVED** [review]
Looks good. A few concerns to be addressed, but overall, it's nearly ready. I'll also want to see the tests passing, but the failures may be unrelated (I haven't investigated yet).

**Comment by jaraco** [comment]
Tests were failing. Fixed in 6f8fa0b.

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:139]: Oh. I notice that `slacker` hasn't been updated in years and has been abandoned. It's probably not wise to add new behavior against a deprecated API. Is there any way to implement this behavior using `self.slack`?

File: pmxbot/slack.py
Comment (jaraco) [inline pmxbot/slack.py:142]: Is the right thing to do on an unsuccessful response to pretend like nothing happened? This `None` value will get cached. My instinct is this condition should trigger an Exception.

File: pmxbot/slack.py
Comment (jaraco) [reply pmxbot/slack.py]: Done in 6dfa674.

File: pmxbot/slack.py
Comment (jaraco) [reply pmxbot/slack.py:87]: Oh, I see you answered the question in the description; "keep it simple". Sounds good. I've looked into the rtm API code and it uses threads to manage its concurrency, so I'm not too worried about concurrency concerns.

File: setup.cfg
Comment (chrisBrookes93) [reply setup.cfg]: Agreed, removed

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: My thinking was that 0.1 was a little low and would mean a lot of loop spinning. I think in reality increasing it to 1 would have little noticeable impact. From what I understand any pmxbot actions generated from direct user interaction (e.g. `!thank`) do not get pushed onto the scheduler queue, they're dealt with when the message event is received. This would only be a slight delay in the actions of other plugins (e.g. standup notifications and webhooks) by up to one second.

However, sticking with the philosophy of minimal changes here, I will change back to `0.1`

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: agreed, actioned

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: There is a valid case where a message doesn't have a name, which is when you're talking directly to pmxbot (an IM)
image

As also seen in the example from the docs: https://api.slack.com/methods/conversations.list#:~:text=%22dGVhbTpDMDYxRkE1UEI%3D%22%0A%20%20%20%20%7D%0A%7D-,Alternate%20response,-Example%20response%20when

In this case I think we'd want to continue with a channel of `None` so that the actions still function, since we'll only need the channel_id.

So `name` isn't always present, but I think its fair to assume that `channel` is always present in the body if the request is successful so I will change to:

`self.slacker.conversations.info(msg['channel']).body['channel'].get('name')`

... but will also adapt so that we're not using `slacker` as you suggested in another comment.

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: I hadn't considered the caching of None, I agree with your approach, I will action.

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: I agree, I will rework this and the other place that was changed to use `slacker`

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py:67]: hmm, apparently not possible as slack validates the function signature:
```
    raise SlackClientError(error)
slack_sdk.errors.SlackClientError: The listener 'handle_message' must accept two args: client, event (actual: self, client, msg)
```

We can't really make `handle_message` static since it uses `self` all over

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: Use of `slacker` removed entirely

File: pmxbot/slack.py
Comment (chrisBrookes93) [reply pmxbot/slack.py]: Using `self.slack.web_client.<endpoint>` to talk to slack does not return a `successful` field. Instead it will raise an exception, so I think this gives us what we want.

**Comment by chrisBrookes93** [comment]
@jaraco Thanks for the review and contributions, I particularly like your implementation for handling paginated data from slack!

I believe I've addressed all of your comments, please take another look when you have capacity.

I'd like to do some final testing tomorrow, so if you review and approve before then, don't merge just yet!

File: pmxbot/slack.py
Comment (chrisBrookes93) [inline pmxbot/slack.py:7]: TODO - tidy up git history before merge

**Comment by chrisBrookes93** [comment]
> @jaraco Thanks for the review and contributions, I particularly like your implementation for handling paginated data from slack!
> 
> I believe I've addressed all of your comments, please take another look when you have capacity.
> 
> I'd like to do some final testing tomorrow, so if you review and approve before then, don't merge just yet!

... All looks good to me!

File: pmxbot/slack.py
Comment (jaraco) [reply pmxbot/slack.py]: Thanks for the investigation and explanation.

**Comment by jaraco** [comment]
Excellent work and great collaboration. Thanks!

**Pull request merged**
